File: Servers/GatewayServer/GatewayServer.cs
using System;
using System.Collections.Generic;
using CommonAPI;
using CommonLibraries;
using NodeJS.HttpModule;
using SocketIOLibrary;
namespace GatewayServer
{
    public class GatewayServer
    {
        public JsDictionary<string, GatewayUserModel> users = new JsDictionary<string, GatewayUserModel>();

        public GatewayServer()
        {
            var app = Http.CreateServer((req, res) => res.End());

            var io = SocketIO.Listen(app);
            var port = 1800 + Math.Truncate((int) ( Math.Random() * 4000 ));

            app.Listen(port);
            io.Set("log level", 1);
            var myName = "Gateway " + Guid.NewGuid();

            io.Sockets.On("connection",
                          new Action<SocketIOConnection>((socket) => {
                                                             GatewayUserModel user = null;
                                                             socket.On("Gateway.Message", new Action<GatewayMessageModel>(data => { }));

                                                             socket.On("Gateway.Login",
                                                                       new Action<GatewayLoginMessageModel>((data) => {
                                                                                                                user = new GatewayUserModel();
                                                                                                                user.Socket = socket;
                                                                                                                user.UserName = data.UserName;
                                                                                                                users[data.UserName] = user;
                                                                                                                socket.Emit("Area.Main.Login.Response", "hi! " + data.UserName);
                                                                                                            }));
                                                             socket.On("disconnect", new Action<string>((string data) => users.Remove(user.UserName)));
                                                         }));
        }

        private static void Main()
        {
            new GatewayServer();
        }

        private void messageReceived(string gateway, UserModel user, string eventChannel, ChannelListenTriggerMessage content)
        {
            if (users.ContainsKey(user.UserName)) {
                var u = users[user.UserName];
                u.Socket.Emit("Client.Message", new SocketClientMessageModel(user, eventChannel, content));
            }
        }
    }
}

File: output/GatewayServer.js
require('mscorlib');
var http = require('http');
var socketio = require('socket.io');
////////////////////////////////////////////////////////////////////////////////
// GatewayServer.GatewayServer
var $GatewayServer_GatewayServer = function() {
	this.users = {};
	var app = http.createServer(function(req, res) {
		res.end();
	});
	var io = socketio.listen(app);
	var port = 1800 + (ss.Int32.trunc(Math.random() * 4000) | 0);
	app.listen(port);
	io.set('log level', 1);
	var myName = 'Gateway ' + CommonLibraries.Guid.newGuid();
	io.sockets.on('connection', Function.mkdel(this, function(socket) {
		var user = null;
		socket.on('Gateway.Message', function(data) {
		});
		socket.on('Gateway.Login', Function.mkdel(this, function(data1) {
			user = CommonLibraries.GatewayUserModel.$ctor();
			user.socket = socket;
			user.userName = data1.userName;
			this.users[data1.userName] = user;
			socket.emit('Area.Main.Login.Response', 'hi! ' + data1.userName);
		}));
		socket.on('disconnect', Function.mkdel(this, function(data2) {
			delete this.users[user.userName];
		}));
	}));
};
$GatewayServer_GatewayServer.prototype = {
	$messageReceived: function(gateway, user, eventChannel, content) {
		if (Object.keyExists(this.users, user.userName)) {
			var u = this.users[user.userName];
			u.socket.emit('Client.Message', new CommonLibraries.SocketClientMessageModel(user, eventChannel, content));
		}
	}
};
$GatewayServer_GatewayServer.$main = function() {
	new $GatewayServer_GatewayServer();
};
Type.registerClass(global, 'GatewayServer.GatewayServer', $GatewayServer_GatewayServer, Object);
$GatewayServer_GatewayServer.$main();
